### PR TITLE
test: use grcov to generate coverage report

### PR DESCRIPTION
**Pull request by quake**
This PR changed the coverage report tool from tarpaulin to [grcov](https://github.com/mozilla/grcov), grcov is more simple to set up and it provides reporting for branches and calls count. Just run `make cov`, a `lcov-unit-test.info` file will be genearted in the root path. You can use some IDE [extension](https://github.com/ryanluker/vscode-coverage-gutters) to view the generated report, for example: 

image

### CODE REVIEW

**Comment by quake** [comment]
according to [grocv doc](https://github.com/mozilla/grcov#grcov-with-travis), `lcov.info` file can be uploaded to codecov.io by one-line:
```
bash <(curl -s https://codecov.io/bash) -f lcov.info
```

I will create a new PR to add this to CI.

File: Makefile [hunk 1]
```diff
@@ -14,12 +14,13 @@ CARGO_TARGET_DIR ?= $(shell pwd)/target
...
-# Tarpaulin only supports x86_64 processors running Linux.
-# https://github.com/xd009642/tarpaulin/issues/161
-# https://github.com/xd009642/tarpaulin/issues/190#issuecomment-473564880
 .PHONY: cov
 cov: ## Run code coverage.
-	RUSTC="$$(pwd)/devtools/cov/rustc-proptest-fix" taskset -c 0 cargo tarpaulin --timeout 300 --exclude-files "*/generated/" "test/*" "*/tests/" --all -v --out Xml
+	rustup component add llvm-tools-preview --toolchain nightly
+	grcov --version || cargo +nightly install grcov
+	RUSTFLAGS="-Zinstrument-coverage" LLVM_PROFILE_FILE="ckb-cov-%p-%m.profraw" cargo +nightly test --all
+	grcov . --binary-path ./target/debug/ -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov-unit-test.info
```
Comment (doitian) [inline Makefile:22]:
```suggestion
	grcov . --binary-path "${CARGO_TARGET_DIR}/debug/" -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov-unit-test.info
```

File: Makefile [hunk 2]
```diff
@@ -44,6 +45,15 @@ integration: submodule-init setup-ckb-test ## Run integration tests in "test" di
 integration-release: submodule-init setup-ckb-test prod
 	RUST_BACKTRACE=1 RUST_LOG=${INTEGRATION_RUST_LOG} test/run.sh --release -- --bin ${CARGO_TARGET_DIR}/release/ckb ${CKB_TEST_ARGS}
 
+.PHONY: integration-cov
+integration-cov: submodule-init setup-ckb-test ## Run integration tests and genearte coverage report.
+	rustup component add llvm-tools-preview --toolchain nightly
+	grcov --version || cargo +nightly install grcov
+	RUSTFLAGS="-Zinstrument-coverage" LLVM_PROFILE_FILE="ckb-cov-%p-%m.profraw" cargo +nightly build --features deadlock_detection
+	RUST_BACKTRACE=1 RUST_LOG=${INTEGRATION_RUST_LOG} test/run.sh -- --bin ${CARGO_TARGET_DIR}/debug/ckb ${CKB_TEST_ARGS}
+	grcov . --binary-path ./target/debug/ -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov-integration-test.info
```
Comment (doitian) [inline Makefile:54]:
```suggestion
	grcov . --binary-path "${CARGO_TARGET_DIR}/debug/" -s . -t lcov --branch --ignore-not-existing --ignore "/*" -o lcov-integration-test.info
```

**Review by doitian — CHANGES_REQUESTED** [review]

**Comment by doitian** [comment]
> according to [grocv doc](https://github.com/mozilla/grcov#grcov-with-travis), `lcov-unit-test.info` file can be uploaded to codecov.io by one-line:
> 
> ```
> bash <(curl -s https://codecov.io/bash) -f lcov-unit-test.info
> ```
> 
> I will create a new PR to add this to CI.

I recommend adding a Makefile target to upload local report to codecov first.

**Comment by doitian** [comment]
> > according to [grocv doc](https://github.com/mozilla/grcov#grcov-with-travis), `lcov-unit-test.info` file can be uploaded to codecov.io by one-line:
> > ```
> > bash <(curl -s https://codecov.io/bash) -f lcov-unit-test.info
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I will create a new PR to add this to CI.
> 
> I recommend adding a Makefile target to upload local report to codecov first.

```
curl -Os https://uploader.codecov.io/latest/linux/codecov

chmod +x codecov
./codecov -t ${CODECOV_TOKEN}
```

**Comment by quake** [comment]
> > I will create a new PR to add this to CI.
> 
> It seems our CI is too poor so that it can't bear that. 😢

I plan to add a schedule action to upload develop branch coverage report only once per day.

**Comment by quake** [comment]
> > > according to [grocv doc](https://github.com/mozilla/grcov#grcov-with-travis), `lcov-unit-test.info` file can be uploaded to codecov.io by one-line:
> > > ```
> > > bash <(curl -s https://codecov.io/bash) -f lcov-unit-test.info
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I will create a new PR to add this to CI.
> > 
> > 
> > I recommend adding a Makefile target to upload local report to codecov first.
> 
> ```
> curl -Os https://uploader.codecov.io/latest/linux/codecov
> 
> chmod +x codecov
> ./codecov -t ${CODECOV_TOKEN}
> ```

I have uploaded it via curl:

```
curl -X POST \
  --data-binary @lcov-unit-test.info \
"https://codecov.io/upload/v2?commit=xxx&token=xxx&flags=unit"

curl -X POST \
  --data-binary @lcov-integration-test.info \
"https://codecov.io/upload/v2?commit=xxx&token=xxx&flags=integration"
```

check out:
https://app.codecov.io/gh/nervosnetwork/ckb

**Comment by quake** [comment]
> > > according to [grocv doc](https://github.com/mozilla/grcov#grcov-with-travis), `lcov-unit-test.info` file can be uploaded to codecov.io by one-line:
> > > ```
> > > bash <(curl -s https://codecov.io/bash) -f lcov-unit-test.info
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I will create a new PR to add this to CI.
> > 
> > 
> > I recommend adding a Makefile target to upload local report to codecov first.
> 
> ```
> curl -Os https://uploader.codecov.io/latest/linux/codecov
> 
> chmod +x codecov
> ./codecov -t ${CODECOV_TOKEN}
> ```

I have uploaded it via curl:

```
curl -X POST \
  --data-binary @lcov-unit-test.info \
"https://codecov.io/upload/v2?commit=xxx&token=xxx&flags=unit"

curl -X POST \
  --data-binary @lcov-integration-test.info \
"https://codecov.io/upload/v2?commit=xxx&token=xxx&flags=integration"
```

check out:
https://app.codecov.io/gh/nervosnetwork/ckb

**Review by doitian — APPROVED** [review]

**Comment by doitian** [comment]
@keroro520 @driftluo please review

**Review by keroro520 — APPROVED** [review]

**Comment by keroro520** [comment]
bors r=doitian,keroro520

**Pull request merged**